tests/php/includes/wc-stock-functions-tests.php

```diff
@@ -0,0 +1,185 @@
+
+	/**
+	 * @var array List of statuses which reduces stock from inventory.
+	 */
+	public $order_stock_reduce_statuses = array(
+		'wc-processing',
+		'wc-completed',
+		'wc-on-hold',
+	);
+
+	/**
+	 * @var array List of statuses which restores stock back into inventory.
+	 */
+	public $order_stock_restore_statuses = array(
+		'wc-cancelled',
+		'wc-pending',
+	);
+
+	/**
+	 * @var array List of statuses which have no impact on inventory.
+	 */
+	public $order_stock_no_effect_statuses = array(
+		'wc-failed',
+		'wc-refunded',
+	);
+
+	/**
+	 * Helper function to simulate creating order from cart.
+	 *
+	 * @param string $status_from Status for the newly created order.
+	 */
+	private function create_order_from_cart_with_status( $status_from ) {
+		$product = WC_Helper_Product::create_simple_product(
+			true,
+			array(
+				'manage_stock'   => true,
+				'stock_quantity' => 10,
+			)
+		);
+		WC()->cart->empty_cart();
+		WC()->cart->add_to_cart( $product->get_id(), 1 );
+		WC()->cart->calculate_totals();
+
+		$checkout = WC_Checkout::instance();
+		$order = new WC_Order();
+		$checkout->set_data_from_cart( $order );
+		$order->set_status( $status_from );
+		$order->save();
+		return $order;
+	}
+
+	/**
+	 * Helper function to change order status and assert product inventory values.
+	 *
+	 * @param string $status_from       Initial status of the order.
+	 * @param string $status_to         Status to transition the order to.
+	 * @param int    $before_transition Inventory value before order status is changed.
+	 * @param int    $after_transition  Inventory value after order status change.
+	 */
+	private function transition_order_status_and_assert_stock_quantity( $status_from, $status_to, $before_transition, $after_transition ) {
+		$order      = $this->create_order_from_cart_with_status( $status_from );
+		$order_item = array_values( $order->get_items( 'line_item' ) )[0];
+
+		$product = new WC_Product( $order_item->get_product_id() );
+		$this->assertEquals( $before_transition, $product->get_stock_quantity() );
+
+		// Changing status from UI also calls this method in metadata save hook. This has impact on stock levels, so simulate it here as well.
+		wc_save_order_items( $order, $order->get_items() );
+
+		$order->set_status( $status_to );
+		$order->save();
+		$product = new WC_Product( $order_item->get_product_id() );
+		$this->assertEquals( $after_transition, $product->get_stock_quantity(), "Stock levels unexpected when transitioning from $status_from to $status_to." );
+	}
+
+	/**
+	 * Test inventory count after order status transtions which reduces stock to another status which also reduces stock.
+	 */
+	public function test_status_transition_stock_reduce_to_stock_reduce() {
+		foreach ( $this->order_stock_reduce_statuses as $order_status_from ) {
+			foreach ( $this->order_stock_reduce_statuses as $order_status_to ) {
+				$this->transition_order_status_and_assert_stock_quantity( $order_status_from, $order_status_to, 9, 9 );
```

Comment:
Looks like adding it in assert messages is a bit confusing because assert can fail whenever expected and actual stock level would not match, which can happen for a lot of reason than just stock being reduced twice (for example, there could be a bug which increases the stock, and assert message will say stock is not reduced twice).

But I agree that its a bit unclear at the moment, so I have added expectations in doc comments for all these tests in https://github.com/woocommerce/woocommerce/pull/28620/commits/5171108967750ad8c6fe7ab87ccd4d6139268578. Let me know if these makes sense.